tests/python_package_test/test_dask.py

```diff
@@ -717,6 +725,235 @@ def test_ranker(output, group, boosting_type, tree_learner, cluster):
+
+                # check that each eval_name and metric exists for all eval sets, allowing for the
+                # case when a worker receives a fully-padded eval_set component which is not evaluated.
+                if evals_result[eval_name] != 'not evaluated':
+                    for metric in eval_metric_names:
+                        assert metric in evals_result[eval_name]
+                        assert metric in best_scores[eval_name]
+                        assert len(evals_result[eval_name][metric]) == fit_trees
+
+
+@pytest.mark.parametrize('task', ['binary-classification', 'regression', 'ranking'])
+def test_eval_set_with_custom_eval_metric(task, cluster):
+    with Client(cluster) as client:
+
+        n_samples = 1000
+        n_eval_samples = int(n_samples * 0.5)
+        chunk_size = 10
+        output = 'array'
+
+        X, y, w, g, dX, dy, dw, dg = _create_data(
+            objective=task,
+            n_samples=n_samples,
+            output=output,
+            chunk_size=chunk_size
+        )
+        _, y_e, _, _, dX_e, dy_e, _, dg_e = _create_data(
+            objective=task,
+            n_samples=n_eval_samples,
+            output=output,
+            chunk_size=chunk_size
+        )
+
+        if task == 'ranking':
+            eval_at = '5,6'
+            eval_metrics = ['ndcg', _constant_metric]
+            eval_metric_names = [f'ndcg@{k}' for k in eval_at.split(',')] + ['constant_metric']
+        else:
+            if task == 'binary-classification':
+                eval_metrics = ['binary_error', 'auc', _constant_metric]
+                eval_metric_names = ['binary_logloss', 'binary_error', 'auc', 'constant_metric']
+            else:
+                eval_metrics = ['l1', _constant_metric]
+                eval_metric_names = ['l2', 'l1', 'constant_metric']
+
+        fit_trees = 50
+        params = {
+            "random_state": 42,
+            "n_estimators": fit_trees,
+            "num_leaves": 2
+        }
+        model_factory = task_to_dask_factory[task]
+        dask_model = model_factory(
+            client=client,
+            **params
+        )
+
+        eval_set = [(dX_e, dy_e)]
+        fit_params = {
+            'X': dX,
+            'y': dy,
+            'eval_set': eval_set,
+            'eval_metric': eval_metrics
+        }
+        if task == 'ranking':
+            fit_params.update(
+                {'group': dg,
+                 'eval_group': [dg_e],
+                 'eval_at': eval_at}
+            )
+
+        dask_model = dask_model.fit(**fit_params)
+
+        eval_name = 'valid_0'
+        evals_result = dask_model.evals_result_
+        assert len(evals_result) == 1
+        assert eval_name in evals_result
+
+        for metric in eval_metric_names:
+            assert metric in evals_result[eval_name]
+            assert len(evals_result[eval_name][metric]) == fit_trees
+
+        assert np.allclose(evals_result[eval_name]['constant_metric'], 0.708)
```

Comment:
We prefer using `numpy.testing` module for asserting arrays.
```suggestion
        np.testing.assert_allclose(evals_result[eval_name]['constant_metric'], 0.708)
```

Reply:
Addressed in 703e6f7b